src/libxtp/davidsonsolver.cc

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2009-2018 The VOTCA Development Team (http://www.votca.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <stdexcept>
+#include <iostream>
+
+#include <votca/xtp/eigen.h>
+#include <votca/xtp/davidsonsolver.h>
+
+
+
+using boost::format;
+using std::flush;
+
+namespace votca { 
+    namespace xtp {
+
+using namespace std;
+
+DavidsonSolver::DavidsonSolver(ctp::Logger &log) : _log(log) { }
+
+void DavidsonSolver::set_correction(std::string method) {
+    if (method == "DPR") this->davidson_correction = CORR::DPR;
+    else if (method == "JACOBI") this->davidson_correction = CORR::JACOBI;
+    else throw std::runtime_error("Not a valid correction method");
+}
+
+void DavidsonSolver::set_jacobi_linsolve(std::string method) {
+    if (method == "CG") this->jacobi_linsolve = LSOLVE::CG;
+    else if (method == "GMRES") this->jacobi_linsolve = LSOLVE::GMRES;
+    else if (method == "LLT") this->jacobi_linsolve = LSOLVE::LLT;   
+    else throw std::runtime_error("Not a valid linsolve method");
+}
+
+Eigen::ArrayXd DavidsonSolver::_sort_index(Eigen::VectorXd &V) const
+{
+    Eigen::ArrayXd idx = Eigen::ArrayXd::LinSpaced(V.rows(),0,V.rows()-1);
+    std::sort(idx.data(),idx.data()+idx.size(),
+              [&](int i1, int i2){return V[i1]<V[i2];});
+    return idx; 
+}
+
+Eigen::MatrixXd DavidsonSolver::_get_initial_eigenvectors(Eigen::VectorXd &d, int size_initial_guess) const
+{
+    Eigen::MatrixXd guess = Eigen::MatrixXd::Zero(d.size(),size_initial_guess);
+    Eigen::ArrayXd idx = DavidsonSolver::_sort_index(d);
```

Comment:
It could be int indeed. It's to define the first guess of the eigenvectors such that they 'target' the lowest diagonal element of the matrix. But it turns out that doesn't help that much in the end. And I don't quite know why ...

Reply:
I've put it as int